src/cli/commands/files/add.js

```diff
@@ -55,44 +55,71 @@ module.exports = {
     const index = inPath.lastIndexOf('/') + 1
 
     utils.getIPFS((err, ipfs) => {
-      if (err) throw err
+      if (err) {
+        throw err
+      }
+
+      // pull-stream -> stream -> pull-stream
+      // is not a good idea, so if we have a pull interface
+      // available, use it
+      let createAddStream = (cb) => {
+        ipfs.files.createAddStream((err, stream) => {
+          cb(err, err ? null : toPull.transform(stream))
+        })
+      }
+
+      if (typeof ipfs.files.createAddPullStream === 'function') {
+        createAddStream = (cb) => {
+          cb(null, ipfs.files.createAddPullStream())
+        }
+      }
+
+      createAddStream((err, addStream) => {
+        if (err) {
+          throw err
+        }
+
+        glob(path.join(inPath, '/**/*'), (err, list) => {
+          if (err) throw err
+          if (list.length === 0) list = [inPath]
 
-      glob(path.join(inPath, '/**/*'), (err, list) => {
-        if (err) throw err
-        if (list.length === 0) list = [inPath]
-
-        pull(
-          zip(
-            pull.values(list),
-            pull(
-              pull.values(list),
-              paramap(fs.stat.bind(fs), 50)
-            )
-          ),
-          pull.map((pair) => ({
-            path: pair[0],
-            isDirectory: pair[1].isDirectory()
-          })),
-          pull.filter((file) => !file.isDirectory),
-          pull.map((file) => ({
-            path: file.path.substring(index, file.path.length),
-            content: pullFile(file.path)
-          })),
-          ipfs.files.createAddPullStream(),
-          pull.map((file) => ({
-            hash: file.hash,
-            path: file.path
-          })),
-          pull.collect((err, added) => {
-            if (err) throw err
-
-            sortBy(added, 'path')
-              .reverse()
-              .map((file) => `added ${file.hash} ${file.path}`)
-              .forEach((msg) => console.log(msg))
-          })
-        )
+          addPipeline(index, addStream, list)
+        })
       })
     })
   }
 }
+
+function addPipeline (index, addStream, list) {
+  pull(
+    zip(
+      pull.values(list),
+      pull(
+        pull.values(list),
+        paramap(fs.stat.bind(fs), 50)
+      )
+    ),
```

Comment:
oh, wanna make a comment on that line, it was clearly not obvious (I would expect that to be called concat)